src/rules/unbalanced-delimiters.ts

```diff
@@ -0,0 +1,132 @@
+import * as parse5 from 'parse5';
+import * as matchers from '../matchers';
+import * as utils from '../utils';
+
+export class UnbalancedDelimiters implements Rule {
+  public async check(document: Document): Promise<Warning[]> {
+    let warnings: Warning[] = [];
+
+    const parsedHtml = document.parsedDocument;
+
+    // This linter is only relevant to html documents, so exit out unless the
+    // provided document is one.
+    if (!(parsedHtml instanceof ParsedHtmlDocument)) {
+      return warnings;
+    }
+
+    // Address binding expressions only for polymer elements.
+    for (const element of document.getByKind('polymer-element')) {
+      console.assert(element instanceof PolymerElement);
+
+      for (const warning of this._warningsForPolymerElement(
+               parsedHtml, element)) {
+        warnings.push(warning);
+      }
+    }
+
+    return warnings;
+  }
+
+  private _warningsForPolymerElement(
+      parsedHtml: ParsedHtmlDocument,
+      element: PolymerElement): Warning[] {
+    let warnings: Warning[] = [];
+
+    if (!element.domModule) {
+      return warnings;
+    }
+
+    const template = dom5.query(element.domModule, matchers.isTemplate);
+    if (!template) {
+      return warnings;
+    }
+
+    const templateContent =
+        parse5.treeAdapters.default.getTemplateContent(template);
+
+    dom5.nodeWalkAll(templateContent, (node: parse5.ASTNode) => {
+
+      if (dom5.isElement(node) && node.attrs.length > 0) {
+        for (const warning of this._warningsForElementAttrs(parsedHtml, node)) {
+          warnings.push(warning);
+        }
+      }
+      if (dom5.isTextNode(node) && typeof node.value === 'string' &&
+          this._isBadBindingExpression(node.value)) {
+        warnings.push({
+          code: 'unbalanced-delimiters',
+          message: `Expression ${node.value} has unbalanced delimiters`,
+          severity: Severity.ERROR,
+          sourceRange: utils.getSourceRangeForTextNode(parsedHtml.url, node)!
+        });
+      }
+
+      return false;
+    });
+    return warnings;
+  }
+
+  private _warningsForElementAttrs(
+      parsedHtml: ParsedHtmlDocument,
+      element: parse5.ASTNode): Warning[] {
+    const warnings: Warning[] = [];
+    for (const attr of element.attrs) {
+      if (this._extractBadBindingExpression(attr.value || '')) {
+        warnings.push({
+          code: 'unbalanced-delimiters',
+          message: `Expression ${attr
+                       .value
+                   } has unbalanced delimiters in attribute ${attr.name}`,
+          severity: Severity.ERROR,
+          // TODO(usergenic): Extend polymer-analyzer's ParsedHtmlDocument to
+          // provide sourceRangeForAttributeValue specifically.
+          sourceRange:
+              parsedHtml.sourceRangeForAttribute(element, attr.name)!
+        });
+      }
+    }
+    return warnings;
+  }
+
+  private _extractBadBindingExpression(text: string): string|undefined {
+    // 4 cases, {{}, {}}, [[], []]
```

Comment:
If there are just four cases, could we check for them individually? It would make for more legible code (and I say this as a generally regex-loving weirdo), and we could provide better warning messages. For me, I think the ideal message here would be:

`Invalid polymer expression delimiters. You put '{{}' are you missing a closing }`

I think it's worthwhile to say that they're polymer expression delimiters (and not like, a missing closing paren _inside_ the polymer expression), and to say what we saw and how to fix.